File: src/typescript/bubbleSort.ts
const bubbleSort = (seq: any[]): void => {
  for (let i = 0; i < seq.length - 1; i++) {
    let isSorted = true;
    for (let j = 0; j < seq.length - 1 - i; j++) {
      if (seq[j] > seq[j+1]) {
        isSorted = false;
        [seq[j], seq[j + 1]] = [seq[j + 1], seq[j]];
      }
    }
    if (isSorted) {
      break;
    }
  }
}


const bubbleSortTest = () => {
  let seq1 = [16, 7, 9, 5, 65, 49, 37, 3, 28, 2, 21, 12, 4];
  let seq1Sorted = Object.assign([], seq1);
  let seq2 = [
    "p", "g", "i", "j", "65", "W", "K", "c", "B", "b", "21", "l", "d"
  ];
  let seq2Sorted = Object.assign([], seq2);
  bubbleSort(seq1Sorted);
  bubbleSort(seq2Sorted);

  console.log(`seq1: ${seq1}\nseq1Sorted: ${seq1Sorted}`);
  console.log(`seq2: ${seq2}\nseq2Sorted: ${seq2Sorted}`);
}

bubbleSortTest();

File: src/rust/src/sort/bubble_sort.rs
pub fn bubble_sort<T>(seq: &mut [T])
where
    T: PartialOrd
{
    for i in 0..seq.len() {
        let mut is_sorted = true;
        for j in 0..seq.len() - 1 - i {
            if seq[j] > seq[j + 1] {
                is_sorted = false;
                seq.swap(j, j + 1);
            }
        }
        if is_sorted {
            break;
        }
    }
}
